core/src/main/scala/akka/persistence/query/EventBySliceEnvelope.scala

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (C) 2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.query
+
+import java.util.Optional
+
+import akka.util.HashCode
+
+object EventBySliceEnvelope {
+  def apply[Event](
+      offset: Offset,
+      persistenceId: String,
+      sequenceNr: Long,
+      event: Event,
+      timestamp: Long,
+      entityType: String,
+      slice: Int): EventBySliceEnvelope[Event] =
+    new EventBySliceEnvelope(offset, persistenceId, sequenceNr, Option(event), timestamp, None, entityType, slice)
+
+  def create[Event](
+      offset: Offset,
+      persistenceId: String,
+      sequenceNr: Long,
+      event: Event,
+      timestamp: Long,
+      entityType: String,
+      slice: Int): EventBySliceEnvelope[Event] =
+    apply(offset, persistenceId, sequenceNr, event, timestamp, entityType, slice)
+
+  def unapply[Event](arg: EventBySliceEnvelope[Event]): Option[(Offset, String, Long, Option[Event], Long)] =
+    Some((arg.offset, arg.persistenceId, arg.sequenceNr, arg.eventOption, arg.timestamp))
+}
+
+/**
+ * Event wrapper adding meta data for the events in the result stream of
+ * [[akka.persistence.query.scaladsl.EventsBySliceQuery]] query, or similar queries.
+ *
+ * If the `event` is not defined it has not been loaded yet. It can be loaded with the
+ * [[akka.persistence.query.scaladsl.LoadEventQuery]].
+ *
+ * The `timestamp` is the time the event was stored, in milliseconds since midnight, January 1, 1970 UTC (same as
+ * `System.currentTimeMillis`).
+ */
+final class EventBySliceEnvelope[Event](
+    val offset: Offset,
+    val persistenceId: String,
+    val sequenceNr: Long,
+    val eventOption: Option[Event],
```

Comment:
Would be benefit of making this a private `OptionVal` and only wrap it in `Option` or `Optional` on user request?

Reply:
Good suggestion, but I don't think it is worth it. OptionVal is internal api (InternalStableApi) so we should not encourage usage of it outside of Akka core. Would require additional overloaded constructors (factory methods) and if conditions in plugins to select constructor. E.g. SerializedJournalRow also has an Option field for the event payload.

These queries are anyway allocating rather many objects so saving one allocation here is not worth much.